astropy/table/table.py

```diff
@@ -3861,21 +3861,27 @@ def _is_mixin_for_table(self, col):
 
     def _convert_col_for_table(self, col):
         if isinstance(col, Column) and getattr(col, 'unit', None) is not None:
-            # What to do with MaskedColumn with units: leave as MaskedColumn or
-            # turn into Quantity and drop mask?  Assuming we have masking support
-            # in Quantity someday, let's drop the mask (consistent with legacy
-            # behavior) but issue a warning.
-            if isinstance(col, MaskedColumn) and np.any(col.mask):
-                warnings.warn("dropping mask in Quantity column '{}': "
-                              "masked Quantity not supported".format(col.info.name))
-
             # We need to turn the column into a quantity, or a subclass
             # identified in the unit (such as u.mag()).
             q_cls = getattr(col.unit, '_quantity_class', Quantity)
-            qcol = q_cls(col.data, col.unit, copy=False)
-            qcol.info = col.info
-            qcol.info.indices = col.info.indices
-            col = qcol
+            try:
+                qcol = q_cls(col.data, col.unit, copy=False)
+            except Exception as exc:
+                warnings.warn(f"column {col.info.name} has a unit but is kept as "
```

Comment:
According to our dev docs these should both be an `AstropyUserWarning`:
```
* For warnings, one should always use ``warnings.warn(message,
  warning_class)``. These get redirected to ``log.warning()`` by default,
  but one can still use the standard warning-catching mechanism and custom
  warning classes. The warning class should be either
  :class:`~astropy.utils.exceptions.AstropyUserWarning` or inherit from it.
```

Reply:
OK, that makes sense. While I'm at it, shall I then also change the warning for masked columns just below to be an `AstropyUserWarning`? (this is what I used as my template...)

Reply:
... changed them both.